### PR TITLE
Tech/introduce-preview-api

### PR DESCRIPTION
**Pull request by ndelangen**
Refactor the preview packages

Introduces a new package called `preview-api` which should be the only package the user would ever want to import in any code that runs on the preview side.

This replaces the need to ever import from `@storbook/addons`, `@storybook/preview-web`, `@storybook/client-api`,`@storybook/store`,`@storybook/core-client`.

All these packages are unified under this 1 package.

Likely these packages listed will be removed at some point in the future, simplifying maintenance & future development. It also means users will have an easier time importing the right api from the right place.

I've updated all our code to use the new package.

This should all be backwards compatible, if a user were to still import `@storybook/client-api` for example they'd see a deprecation warning in the console of their browser, but everything should work.

Some of the code of these packages has actually moved, that includes these 4 packages: `client-api`, `core-client`, `preview-web` and `store`. Other packages still have their code in the original place, and the `preview-api` package simply re-exports some or all exports from the original.

The intent is to move all code for the preview into this package when it's not shared anywhere else (runs in the preview only). I decided to make this PR with 4 packages moved, to show it's possible, demonstrate backwards compatibility, and get approval from the other maintainers / rest of the team.

An important step I'm still keen on making is is to split `@storybook/addons` into 2 packages, 1 for the preview (code should be moved to `@storybook/preview-api` and the other part for code that runs in the manager (mostly for registering addon panels/etc.) into the `@storybook/api` package... which I intent to rename to `@storybook/manager-api`.
 
Diagram of packages (note that `manager-api` is currently called `api`, and this is a proposed change not included in this PR)
Untitled Diagram drawio (7)

This is what prebundling really means:
Untitled Diagram drawio (9)

At the builder level we break the dependency graph into pieces. The are pieces that are stories, and there are pieces that is storybook's runtime code. By breaking the storybook runtime code away from the work that the builder has to do, we gain better performance, because there is less work for the builder to do. The user never has their builder compile storybook's code.

SO the package re-architecture (right now) looks like this:
Untitled Diagram drawio (6)
I've moved code of the 4 packages listed TO preview-api, and for backwards compatibility those packages re-export from `preview-api`.
`@storybook/addons` is currently doing something slightly different, because of it's dual use in both the preview and manager (which is a problem) that I will still do. Possibly as part of this PR.

### CODE REVIEW

File: code/addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts [hunk 1]
```diff
@@ -35,8 +35,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts:39]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/html/loader.ts [hunk 2]
```diff
@@ -12,8 +12,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/html/loader.ts:16]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/preact/loader.ts [hunk 3]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/preact/loader.ts:22]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/rax/loader.ts [hunk 4]
```diff
@@ -13,8 +13,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/rax/loader.ts:17]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/react/loader.ts [hunk 5]
```diff
@@ -10,8 +10,8 @@ function test(options: StoryshotsOptions): boolean {
 function load(options: StoryshotsOptions) {
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/react/loader.ts:14]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/riot/loader.ts [hunk 6]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/riot/loader.ts:22]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/svelte/loader.ts [hunk 7]
```diff
@@ -15,8 +15,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/svelte/loader.ts:19]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/vue/loader.ts [hunk 8]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/vue/loader.ts:22]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/vue3/loader.ts [hunk 9]
```diff
@@ -13,8 +13,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/vue3/loader.ts:17]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/addons/storyshots/storyshots-core/src/frameworks/web-components/loader.ts [hunk 10]
```diff
@@ -12,8 +12,8 @@ function load(options: StoryshotsOptions) {
 
   let mockStartedAPI: any;
 
-  jest.mock('@storybook/core-client', () => {
-    const coreClientAPI = jest.requireActual('@storybook/core-client');
+  jest.mock('@storybook/preview-api', () => {
+    const coreClientAPI = jest.requireActual('@storybook/preview-api');
```
Comment (valentinpalkovic) [inline code/addons/storyshots/storyshots-core/src/frameworks/web-components/loader.ts:16]:
```suggestion
    const previewAPI = jest.requireActual('@storybook/preview-api');
```

File: code/frameworks/nextjs/package.json [hunk 11]
```diff
@@ -59,13 +59,15 @@
     "prep": "../../../scripts/prepare/bundle.ts"
   },
   "dependencies": {
+    "@babel/preset-typescript": "^7.18.6",
```
Comment (valentinpalkovic) [inline code/frameworks/nextjs/package.json:62]: Was this addition intended?

File: code/frameworks/vue3-vite/src/index.ts [hunk 12]
```diff
@@ -1,13 +1,8 @@
...
-const { ClientApi } = clientApi;
 export { createChannel as createPostMessageChannel } from '@storybook/channel-postmessage';
 export { createChannel as createWebSocketChannel } from '@storybook/channel-websocket';
-export { addons } from '@storybook/addons';
-export { composeConfigs, PreviewWeb } from '@storybook/preview-web';
-export { clientApi, ClientApi };
+
+export { addons } from '@storybook/preview-api';
+export { composeConfigs, PreviewWeb } from '@storybook/preview-api';
+export { ClientApi } from '@storybook/preview-api';
```
Comment (valentinpalkovic) [inline code/frameworks/vue3-vite/src/index.ts:6]: Move to one export instead of having three

File: code/frameworks/web-components-vite/src/index.ts [hunk 13]
```diff
@@ -1,13 +1,8 @@
...
-const { ClientApi } = clientApi;
 export { createChannel as createPostMessageChannel } from '@storybook/channel-postmessage';
 export { createChannel as createWebSocketChannel } from '@storybook/channel-websocket';
-export { addons } from '@storybook/addons';
-export { composeConfigs, PreviewWeb } from '@storybook/preview-web';
-export { clientApi, ClientApi };
+
+export { addons } from '@storybook/preview-api';
+export { composeConfigs, PreviewWeb } from '@storybook/preview-api';
+export { ClientApi } from '@storybook/preview-api';
```
Comment (valentinpalkovic) [inline code/frameworks/web-components-vite/src/index.ts:6]: Move to one export instead of having three

File: code/lib/preview-api/README.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+1. Read and update the URL (via the URL Store)
+
+2. Listen to instructions on the channel and emit events as things occur.
+
+3. Render the current selection to the web view in either story or docs mode.
+
+## V7 Store vs Legacy (V6)
+
+The story store is designed to load stories 'on demand', and will operate in this fashion if the `storyStoreV7` feature is enabled.
```
Comment (valentinpalkovic) [inline code/lib/preview-api/README.md:15]: Do we have a unified way of typing "story store"? So is it "Story Store", "story store" or "StoryStore"? Minor, but helps to keep domain understanding.

File: code/lib/preview-api/src/modules/preview-web/docs-context/DocsContextProps.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+import type { DocsContextProps } from '@storybook/types';
+
+export { DocsContextProps };
```
Comment (valentinpalkovic) [inline code/lib/preview-api/src/modules/preview-web/docs-context/DocsContextProps.ts:3]: For curiosity. Is a re-export for types possible?

```suggestion
export type { DocsContextProps } from '@storybook/types';
```

File: code/lib/preview-api/src/modules/preview-web/docs-context/DocsRenderFunction.ts [hunk 16]
```diff
@@ -0,0 +1,3 @@
+import type { DocsRenderFunction } from '@storybook/types';
+
+export type { DocsRenderFunction };
```
Comment (valentinpalkovic) [inline code/lib/preview-api/src/modules/preview-web/docs-context/DocsRenderFunction.ts:3]: For curiosity. Is a re-export for types possible?

```suggestion
export type { DocsRenderFunction } from '@storybook/types';
```

File: code/lib/preview-api/src/modules/store/decorators.test.ts [hunk 17]
```diff
@@ -92,7 +92,7 @@ describe('client-api.decorators', () => {
   // the same story twice at the same time.
   it('does not interleave contexts if two decorated stories are call simultaneously', async () => {
     const contexts: StoryContext[] = [];
-    let resolve;
+    let resolve: any;
```
Comment (valentinpalkovic) [inline code/lib/preview-api/src/modules/store/decorators.test.ts:95]:
```suggestion
    let resolve: (value: unknown) => void;
```

File: scripts/prepare/bundle.ts [hunk 18]
```diff
@@ -92,7 +92,7 @@ const run = async ({ cwd, flags }: { cwd: string; flags: string[] }) => {
         c.platform = platform || 'browser';
         c.legalComments = 'none';
         c.minifyWhitespace = optimized;
-        c.minifyIdentifiers = optimized;
+        c.minifyIdentifiers = false;
```
Comment (valentinpalkovic) [inline scripts/prepare/bundle.ts:95]: Could you leave a comment, why this change is needed?

**Comment by tmeasday** [comment]
@ndelangen can all these nice pictures end up in a README or something please?

File: code/lib/preview-api/README.md (hunk 14 above)
Comment (tmeasday) [reply code/lib/preview-api/README.md:15]: I think "Story Store" although I'm not sure we mention it in prose too much.

@ndelangen maybe we should make a ticket to clean up some of these READMEs.

**Review by tmeasday — APPROVED** [review]
Love it @ndelangen !

File: scripts/prepare/bundle.ts (hunk 18 above)
Comment (ndelangen) [reply scripts/prepare/bundle.ts:95]: Yes!

See here: https://circleci.com/gh/storybookjs/storybook/444855

When we build our code for `dist`, we used to compress our variable names.
Turns out when generating variables names like this `_2` is a valid identifier.
But when you try and create a function with that name using `eval` or `new Function()` it throws.

I decided that not compressing variable names would probably be the right choice here, considering this might make debugging way way easier too.

File: code/lib/preview-api/src/modules/preview-web/docs-context/DocsRenderFunction.ts (hunk 16 above)
Comment (ndelangen) [reply code/lib/preview-api/src/modules/preview-web/docs-context/DocsRenderFunction.ts:3]: Yeah, I'm pretty sure that's possible.

File: code/lib/preview-api/README.md (hunk 14 above)
Comment (ndelangen) [reply code/lib/preview-api/README.md:15]: Agreed, This readme, needs a thorough rework, as @tmeasday also points out.

File: code/frameworks/nextjs/package.json (hunk 11 above)
Comment (ndelangen) [reply code/frameworks/nextjs/package.json:62]: doubtful, likely a merge conflict resolved incorrectly

Comment (ndelangen) [inline code/frameworks/nextjs/package.json:62]:
```suggestion
```

File: code/addons/storyshots/storyshots-core/src/frameworks/vue3/loader.ts [hunk 19]
```diff
@@ -13,8 +13,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/vue3/loader.ts:20]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/vue/loader.ts [hunk 20]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/vue/loader.ts:25]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/svelte/loader.ts [hunk 21]
```diff
@@ -15,8 +15,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/svelte/loader.ts:22]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/riot/loader.ts [hunk 22]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/riot/loader.ts:25]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/react/loader.ts [hunk 23]
```diff
@@ -10,8 +10,8 @@ function test(options: StoryshotsOptions): boolean {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/react/loader.ts:17]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/rax/loader.ts [hunk 24]
```diff
@@ -13,8 +13,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/rax/loader.ts:20]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/preact/loader.ts [hunk 25]
```diff
@@ -18,8 +18,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/preact/loader.ts:25]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/html/loader.ts [hunk 26]
```diff
@@ -12,8 +12,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/html/loader.ts:19]:
```suggestion
      ...previewAPI,
```

File: code/addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts [hunk 27]
```diff
@@ -35,8 +35,8 @@ function load(options: StoryshotsOptions) {
...
 
     return {
       ...coreClientAPI,
```
Comment (ndelangen) [inline code/addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts:42]:
```suggestion
      ...previewAPI,
```

**Review by valentinpalkovic — APPROVED** [review]

File: code/lib/preview-api/README.md (hunk 14 above)
Comment (ndelangen) [reply code/lib/preview-api/README.md]: I've moved the README files a bit, so they are in 1 place. They'll likely need extensive work to make proper sense again. I was hoping to do that in a follow-up PR..
